File: frontend/src/pages/api/djs/[id]/library.js
import fs from 'fs';
import path from 'path';
import { resolveDjScript } from '../../../../../lib/dj-scripts';

const AUDIO_EXTS = new Set(['.mp3', '.wav', '.ogg', '.flac', '.aac', '.m4a']);

function normalizeFileName(fileName) {
  return path.basename(fileName.toString());
}

function safeId(value) {
  return value.toString().trim().replace(/\s+/g, '-').replace(/[^a-zA-Z0-9-_\.]/g, '').slice(0, 200);
}

function parseJsonFile(filePath) {
  try {
    return JSON.parse(fs.readFileSync(filePath, 'utf-8')) || {};
  } catch {
    return {};
  }
}

function getAlternateDjRoot(resolved) {
  const scriptDir = path.dirname(resolved.scriptPath);
  return path.resolve(scriptDir, '..', 'djs', resolved.folderName);
}

function getCandidateSongDirs(resolved) {
  const alternateRoot = getAlternateDjRoot(resolved);
  return [...new Set([
    path.join(resolved.folder, 'musica', 'canciones'),
    path.join(resolved.folder, 'canciones'),
    resolved.folder,
    path.join(alternateRoot, 'musica', 'canciones'),
    path.join(alternateRoot, 'canciones'),
    alternateRoot,
  ])];
}

function getCandidateJsonDirs(resolved) {
  const alternateRoot = getAlternateDjRoot(resolved);
  return [...new Set([
    path.join(resolved.folder, 'musica', 'json'),
    path.join(resolved.folder, 'json'),
    path.join(alternateRoot, 'musica', 'json'),
    path.join(alternateRoot, 'json'),
  ])];
}

function buildAudioUrl(djId, fileName) {
  return `/api/djs/${encodeURIComponent(djId)}/audio?file=${encodeURIComponent(fileName)}`;
}

export default async function handler(req, res) {
  if (req.method !== 'GET') {
    return res.status(405).json({ error: 'Método no permitido' });
  }

  const rawId = Array.isArray(req.query.id) ? req.query.id[0] : req.query.id;
  if (!rawId) {
    return res.status(400).json({ error: 'ID de DJ no proporcionado' });
  }

  const resolved = resolveDjScript(rawId.toString());
  if (!resolved) {
    return res.status(404).json({ error: 'DJ no encontrado' });
  }

  const songDirs = getCandidateSongDirs(resolved);
  const jsonDirs = getCandidateJsonDirs(resolved);
  let tracks = [];

  try {
    for (const songsDir of songDirs) {
      if (!fs.existsSync(songsDir)) continue;
      const entries = await fs.promises.readdir(songsDir, { withFileTypes: true });
      for (const entry of entries) {
        if (!entry.isFile()) continue;
        const fileName = normalizeFileName(entry.name);
        const ext = path.extname(fileName).toLowerCase();
        if (!AUDIO_EXTS.has(ext)) continue;

        const stem = path.basename(fileName, ext);
        let meta = {};
        for (const jsonDir of jsonDirs) {
          const metaPath = path.join(jsonDir, `${stem}.json`);
          if (fs.existsSync(metaPath)) {
            meta = parseJsonFile(metaPath);
            break;
          }
        }

        const duration = Number(meta.duracion_segundos || meta.duration || 0) || 0;
        const title = meta.titulo || meta.title || stem;
        const artist = meta.artist || meta.artista || meta.nombre_artistico || undefined;

        tracks.push({
          id: `${safeId(resolved.folderName)}-${safeId(stem)}`,
          title,
          artist,
          url: buildAudioUrl(rawId.toString(), fileName),
          duration,
          file: fileName,
          name: stem,
          bpm: Number(meta.bpm || 0) || 0,
          energia: Number(meta.energia || 50) || 50,
          key: meta.key || '',
          estilo: meta.estilo || '',
          duracion_segundos: duration,
          puede_salir: meta.puede_salir ?? null,
          puede_empezar_mezcla: meta.puede_empezar_mezcla ?? null,
          debe_sonar_sola: meta.debe_sonar_sola ?? null,
          intro_fin: meta.intro_fin ?? null,
          tiene_voz_inicio: meta.tiene_voz_inicio || false,
          beat_times: Array.isArray(meta.beat_times) ? meta.beat_times : [],
          energia_por_segundo: Array.isArray(meta.energia_por_segundo) ? meta.energia_por_segundo : [],
          start_position: Number(meta.puede_empezar_mezcla > 0 ? meta.puede_empezar_mezcla : meta.intro_fin || 0) || 0,
        });
      }
      if (tracks.length > 0) break;
    }

    tracks = tracks.sort((a, b) => a.title.localeCompare(b.title, undefined, { sensitivity: 'base' }));
    res.setHeader('Cache-Control', 'no-store');
    return res.status(200).json(tracks);
  } catch (error) {
    console.error('[library] Error leyendo la librería del DJ:', error);
    return res.status(500).json({ error: 'Error interno leyendo la librería de canciones', details: error.message });
  }
}

File: DJ's/core/library.py
"""Library loader - read songs and metadata from directories."""
import json
from pathlib import Path


def load_library(songs_dir: Path, json_dir: Path) -> list[dict]:
    """Load music library from songs_dir and metadata from json_dir.
    
    Args:
        songs_dir: Path to folder containing MP3/WAV files
        json_dir: Path to folder containing .json metadata files
        
    Returns:
        List of track dicts with metadata
    """
    tracks = []
    exts = {".mp3", ".wav", ".ogg", ".flac", ".aac", ".m4a"}
    
    if not songs_dir.exists():
        return tracks
    
    for f in sorted(songs_dir.iterdir()):
        if f.suffix.lower() not in exts:
            continue
        
        meta = {}
        jp = json_dir / f"{f.stem}.json"
        if jp.exists():
            try:
                raw = json.loads(jp.read_text(encoding="utf-8"))
                meta = {k: v for k, v in raw.items() if not k.startswith("_") and v is not None}
            except:
                pass
        
        # Determine start_position: prefer puede_empezar_mezcla, fallback to intro_fin
        pem = meta.get("puede_empezar_mezcla")
        intro_fin = meta.get("intro_fin")
        if pem and float(pem) > 0:
            start_position = float(pem)
        elif intro_fin and float(intro_fin) > 0:
            start_position = float(intro_fin)
        else:
            start_position = 0.0
        
        tracks.append({
            "name": f.stem,
            "file": f.name,
            "bpm": meta.get("bpm", 0),
            "energia": meta.get("energia", 50),
            "key": meta.get("key", ""),
            "estilo": meta.get("estilo", ""),
            "duracion_segundos": meta.get("duracion_segundos", 0),
            "puede_salir": meta.get("puede_salir"),
            "puede_empezar_mezcla": meta.get("puede_empezar_mezcla"),
            "debe_sonar_sola": meta.get("debe_sonar_sola"),
            "intro_fin": meta.get("intro_fin"),
            "tiene_voz_inicio": meta.get("tiene_voz_inicio", False),
            "beat_times": meta.get("beat_times", []),
            "energia_por_segundo": meta.get("energia_por_segundo", []),
            "start_position": start_position,
        })
    
    return tracks
